readme.txt
=== Loop Post Navigation Links ===
Contributors: coffee2code
Donate link: https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6ARCFJ9TX3522
Tags: posts, navigation, links, next, previous, portfolio, previous_post_link, next_post_link, coffee2code
License: GPLv2 or later
License URI: https://www.gnu.org/licenses/gpl-2.0.html
Requires at least: 4.9
Tested up to: 5.5
Stable tag: 3.0.1

Template tags (for use in single.php) to create post navigation loop (previous to first post is last post; next/after last post is first post).


== Description ==

This plugin provides two template tags for use in single.php to create a post navigation loop, whereby previous to the first post is the last post, and after the last post is first post. Basically, when you're on the last post and you click to go to the next post, the link takes you to the first post. Likewise, if you're on the first post and click to go to the previous post, the link takes you to the last post.

The function `c2c_next_or_loop_post_link()` is identical to WordPress's `next_post_link()` in every way except when called on the last post in the navigation sequence, in which case it links back to the first post in the navigation sequence.

The function `c2c_previous_or_loop_post_link()` is identical to WordPress's `previous_post_link()` in every way except when called on the first post in the navigation sequence, in which case it links back to the last post in the navigation sequence.

Useful for providing a looping link of posts, such as for a portfolio, or to continually present pertinent posts for visitors to continue reading.

If you are interested in getting the post itself and not just a link to the post, you can use the `c2c_get_next_or_loop_post()` and `c2c_get_previous_or_loop_post()` functions. If you just want the URL to the post, you can use `c2c_get_next_or_loop_post_url()` and `c2c_get_previous_or_loop_post_url()`.

Links: [Plugin Homepage](https://coffee2code.com/wp-plugins/loop-post-navigation-links/) | [Plugin Directory Page](https://wordpress.org/plugins/loop-post-navigation-links/) | [GitHub](https://github.com/coffee2code/loop-post-navigation-links/) | [Author Homepage](https://coffee2code.com)


== Installation ==

1. Install via the built-in WordPress plugin installer. Or install the plugin code inside the plugins directory for your site (typically `wp-content/plugins/`).
2. Activate the plugin through the 'Plugins' admin menu in WordPress.
3. Use `c2c_next_or_loop_post_link()` template tag instead of `next_post_link()`, and/or `c2c_previous_or_loop_post_link()` template tag instead of `previous_post_link()`, in your single-post template (e.g. single.php).


== Template Tags ==

The plugin provides four template tags for use in your single-post theme templates.

= Functions =

* `function c2c_next_or_loop_post_link( $format='%link &raquo;', $link='%title', $in_same_term = false, $excluded_terms = '', $taxonomy = 'category' )`
Like WordPress's `next_post_link()`, this function displays a link to the next chronological post (among all published posts, those in the same category, or those not in certain categories). Unlink `next_post_link()`, when on the last post in the sequence this function will link back to the first post in the sequence, creating a circular loop.

* `function c2c_get_next_or_loop_post_link( $format='%link &raquo;', $link='%title', $in_same_term = false, $excluded_terms = '', $taxonomy = 'category' )`
Like `c2c_next_or_loop_post_link(), but returns the value without echoing it.

* `function c2c_previous_or_loop_post_link( $format='&laquo; %link', $link='%title', $in_same_term = false, $excluded_terms = '', $taxonomy = 'category' )`
Like WordPress's `previous_post_link()`, this function displays a link to the previous chronological post (among all published posts, those in the same category, or those not in certain categories). Unlink `previous_post_link()`, when on the first post in the sequence this function will link to the last post in the sequence, creating a circular loop.

* `function c2c_get_previous_or_loop_post_link( $format='&laquo; %link', $link='%title', $in_same_term = false, $excluded_terms = '', $taxonomy = 'category' )`
Like `c2c_get_previous_or_loop_post_link(), but returns the value without echoing it.

* `function c2c_get_next_or_loop_post( $in_same_term = false, $excluded_terms = '', $taxonomy = 'category' )`
Like WordPress's `get_adjacent_post()` when used to find the next post, except when on the last post in the sequence this function will return the first post in the sequence, creating a circular loop.

* `function c2c_get_previous_or_loop_post( $in_same_term = false, $excluded_terms = '', $taxonomy = 'category' )`
Like WordPress's `get_adjacent_post()` when used to find the previous post, except when on the first post in the sequence this function will return the last post in the sequence, creating a circular loop.

* `function c2c_get_next_or_loop_post_url( $in_same_term = false, $excluded_terms = '', $taxonomy = 'category' )`
Returns the URL for the next post or the post at the beginning of the series.

* `function c2c_get_previous_or_loop_post_url( $in_same_term = false, $excluded_terms = '', $taxonomy = 'category' )`
Returns the URL for the previous post or the post at the end of the series.

= Arguments =

* `$format`
(optional) A percent-substitution string indicating the format of the entire output string. Use <code>%link</code> to represent the next/previous post being linked, or <code>%title</code> to represent the title of the next/previous post.

* `$link`
(optional) A percent-substitution string indicating the format of the link itself that gets created for the next/previous post. Use <code>%link</code> to represent the next/previous post being linked, or <code>%title</code> to represent the title of the next/previous post.

* `$in_same_term`
(optional) A boolean value (either true or false) indicating if the next/previous post should be in the current post's same taxonomy term.

* `$excluded_terms`
(optional) An array or comma-separated string of category or term IDs to which posts cannot belong.

* `$taxonomy`
(optional) Taxonomy, if $in_same_term is true. Default 'category'.

== Examples ==

`<div class="loop-navigation">
	<div class="alignleft"><?php c2c_previous_or_loop_post_link(); ?></div>
	<div class="alignright"><?php c2c_next_or_loop_post_link(); ?></div>
</div>`


== Hooks ==

The plugin is further customizable via eleven hooks. Typically, code making use of hooks should ideally be put into a mu-plugin, a site-specific plugin (which is beyond the scope of this readme to explain), or in theme template files. Many of these filters are likely only of interest to advanced users able to code.

**c2c_previous_or_loop_post_link_output, c2c_next_or_loop_post_link_output (filters)**

The 'c2c_previous_or_loop_post_link_output' and 'c2c_next_or_loop_post_link_output' filters allow you to customize the link markup generated for previous and next looping links, respectively.

Arguments:

* $format         (string):       Link anchor format.
* $link           (string):       Link permalink format.
* $in_same_term   (bool):         Optional. Whether link should be in a same taxonomy term. Default is false.
* $excluded_terms (array|string): Optional. Array or comma-separated list of excluded term IDs. Default is ''.
* $previous       (bool):         Optional. Whether to display link to previous or next post. Default is true.
* $taxonomy       (string):       Optional. Taxonomy, if $in_same_term is true. Default 'category'.
* $adjacent       (string):       Whether the post is previous or next.

Example:

  `<?php
    // Prepend "Prev:" to previous link markup.
    function my_custom_previous_or_loop_link_output( $output, $format, $link, $post, $in_same_term, $excluded_terms, $taxonomy ) {
      return 'Prev: ' . $output;
    }
    add_filter( 'c2c_previous_or_loop_post_link_output', 'my_custom_previous_or_loop_link_output', 10, 4 );
  ?>`

**c2c_previous_or_loop_post_link_get, c2c_next_or_loop_post_link_get (filters)**

The 'c2c_previous_or_loop_post_link_get' and 'c2c_next_or_loop_post_link_get' filters allow you to customize the link markups generated for previous and next looping links, respectively, but in the non-echoing functions.

Arguments:

* $output         (string):       The adjacent post link.
* $format         (string):       Link anchor format.
* $link           (string):       Link permalink format.
* $post           (WP_Post):      The adjacent post.
* $in_same_term   (bool):         Optional. Whether link should be in a same taxonomy term. Default is false.
* $excluded_terms (array|string): Optional. Array or comma-separated list of excluded term IDs. Default is ''.
* $previous       (bool):         Optional. Whether to display link to previous or next post. Default is true.
* $taxonomy       (string):       Optional. Taxonomy, if $in_same_term is true. Default 'category'.
* $adjacent       (string):       Whether the post is previous or next.

**c2c_previous_or_loop_post_link, c2c_next_or_loop_post_link, c2c_get_previous_or_loop_post_link, c2c_get_next_or_loop_post_link, c2c_get_adjacent_or_loop_post, c2c_get_previous_or_loop_post, c2c_get_previous_or_loop_post (actions)**

The 'c2c_previous_or_loop_post_link' and 'c2c_next_or_loop_post_link' actions allow you to use an alternative approach to safely invoke `c2c_previous_or_loop_post_link()` and `c2c_next_or_loop_post_link()`, respectively, in such a way that if the plugin were deactivated or deleted, then your calls to the functions won't cause errors in your site. The 'c2c_get_previous_or_loop_post_link' and 'c2c_get_next_or_loop_post_link' filters do the same for the non-echoing `c2c_previous_or_loop_post_link()` and `c2c_next_or_loop_post_link()`.

Arguments:

* Same as for for `c2c_previous_or_loop_post_link()` and `c2c_next_or_loop_post_link()`

Example:

Instead of:


CHANGELOG.md
# Changelog

## _(in-progress)_
* Change: Unset loop flag in class just after it's used to ensure it gets reset
* Change: Escape URL for post before being output in link (hardening)
* Change: Restructure unit test file structure
    * New: Create new subdirectory `phpunit/` to house all files related to unit testing
    * Change: Move `bin/` to `phpunit/bin/`
    * Change: Move `tests/bootstrap.php` to `phpunit/`
    * Change: Move `tests/` to `phpunit/tests/`
    * Change: Rename `phpunit.xml` to `phpunit.xml.dist` per best practices
* Change: Note compatibility through WP 5.5+
* Unit tests:
    * New: Add test for hook registrations
    * New: Add tests for `modify_nextprevious_post_where()`

## 3.0.1 _(2020-05-10)_
* Change: Use HTTPS for link to WP SVN repository in bin script for configuring unit tests
* Change: Note compatibility through WP 5.4+
* Change: Update links to coffee2code.com to be HTTPS
* Change: Unit tests: Remove unnecessary unregistering of hooks

## 3.0 _(2020-01-28)_

### Highlights:

Significant update after a long hiatus! Adds functions for getting the URL of the previous/next post, modernizes unit tests, adds CHANGELOG.md, adds plugin to GitHub, changes plugin initialization, updates compatibility to be for WP 4.9-5.3+, and many other documentation and behind-the-scenes changes.

Note that there is an incompatible fix for missing and incorrect arguments for the `c2c_{$adjacent}_or_loop_post_link_output` filter. This change won't affect you unless you have custom code making use of the filter, which is unlikely for almost all users.

### Details:

* New: Add `c2c_get_next_or_loop_post_url()` and `c2c_get_previous_or_loop_post_url()` to get URL of adjacent/looped post
* New: Add CHANGELOG.md file and move all but most recent changelog entries into it
* New: Add README.md file
* New: Add TODO.md and move existing TODO list from top of main plugin file into it (and add more items to the list)
* New: Add inline documentation for hooks
* New: Add GitHub link to readme.txt
* Fix: Add missing argument `$previous` and remove argument `$post` from `c2c_{$adjacent}_or_loop_post_link_output` hook invocation
* Change: Add `$adjacent` argument to a number of filters, to sync with WP core
    * Adds to `{$adjacent}_post_link`, `{$adjacent}_or_loop_post_link`, `c2c_{$adjacent}_or_loop_post_link_get`, and `c2c_{$adjacent}_or_loop_post_link_output`
* Change: Use `c2c_get_adjacent_or_loop_post()` to obtain post, rather than duplicating its functionality in `c2c_get_adjacent_or_loop_post_link()`
* Change: Remove unnecessary (and incorrect) determination of adjacent post in `c2c_adjacent_or_loop_post_link()`
* Change: Use a different variable name to avoid changing variable sent as function argument and later passed as argument to filter
* Change: Initialize plugin on 'plugins_loaded' action instead of on load
* Change: Use `apply_filters_deprecated()` to formally deprecate the `{$adjacent}_or_loop_post_link` filter
* Change: Update code formatting to match modern WordPress standards
* Change: Remove `load_textdomain()` and just load the textdomain within `init()`
* Unit tests:
    * Change: Change `expected()` to optionally not include arrow quotes
    * Change: Update unit test install script and bootstrap to use latest WP unit test repo
    * Change: Enable more error output for unit tests
    * Change: Comment out unit tests that weren't actually testing anything
    * Fix: Don't declare `$posts` as being static since it's never referenced as if it was
    * New: Add tests for hooks
* Change: Note compatibility through WP 5.3+
* Change: Drop compatibility with versions of WP older than 4.9
* Change: Include documentation on filter arguments
* Change: Tweak formatting for installation instructions
* Change: Remove documentation for non-existent parameters of `modify_nextprevious_post_where()`
* Change: Remove unnecessary `echo` in code example
* Change: Modify formatting of hook name in readme to prevent being uppercased when shown in the Plugin Directory
* Change: Rename readme.txt section from 'Filters' to 'Hooks' and provide a better section intro
* Change: Update License URI to be HTTPS
* Change: Update copyright date (2020)
* Change: Update installation instruction to prefer built-in installer over .zip file

## 2.6.1 _(2016-03-10)_
* New: Add support for language packs:
    * Define 'Text Domain' header attribute.
    * Load textdomain.
* New: Add LICENSE file.
* New: Create empty index.php to prevent files from being listed if web server has enabled directory listings.
* Change: Note compatibility through WP 4.4+.
* Change: Explicitly declare methods in unit tests as public.
* Change: Update copyright date (2016).

## 2.6 _(2015-07-14)_
* Feature: Add new template tags for getting the adjacent or looped post object:
  * `c2c_get_next_or_loop_post`
  * `c2c_get_previous_or_loop_post`
  * `c2c_get_adjacent_or_loop_post`

loop-post-navigation-links.php
<?php
/**
 * Plugin Name: Loop Post Navigation Links
 * Version:     3.0.1
 * Plugin URI:  https://coffee2code.com/wp-plugins/loop-post-navigation-links/
 * Author:      Scott Reilly
 * Author URI:  https://coffee2code.com/
 * Text Domain: loop-post-navigation-links
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 * Description: Template tags (for single.php) to create post navigation loop (previous to first post is last post; next/after last post is first post).
 *
 * Compatible with WordPress 4.9 through 5.5.
 *
 * =>> Read the accompanying readme.txt file for instructions and documentation.
 * =>> Also, visit the plugin's homepage for additional information and updates.
 * =>> Or visit: https://wordpress.org/plugins/loop-post-navigation-links/
 *
 * @package Loop_Post_Navigation_Links
 * @author  Scott Reilly
 * @version 3.0.1
 */

/*
	Copyright (c) 2008-2020 by Scott Reilly (aka coffee2code)

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

defined( 'ABSPATH' ) or die();

if ( ! class_exists( 'c2c_LoopPostNavigationLinks' ) ) :

class c2c_LoopPostNavigationLinks {

	/**
	 * Flag to indicate if loop navigation is currently enanbled.
	 *
	 * @since 2.6
	 * @var bool
	 */
	public static $loop_navigation_find = false;

	/**
	 * Returns version of the plugin.
	 *
	 * @since 2.6
	 */
	public static function version() {
		return '3.0.1';
	}

	/**
	 * Class constructor: initializes class variables and adds actions and filters.
	 *
	 * @since 2.6
	 */
	public static function init() {
		// Load plugin textdomain.
		load_plugin_textdomain( 'loop-post-navigation-links' );

		/*
		 * Register actions to filter WHERE clause when previous or next post query is being processed.
		 */
		add_filter( 'get_next_post_where',     array( __CLASS__, 'modify_nextprevious_post_where' ) );
		add_filter( 'get_previous_post_where', array( __CLASS__, 'modify_nextprevious_post_where' ) );
	}

	/**
	 * Modifies the SQL WHERE clause used by WordPress when getting a previous/next post to accommodate looping navigation.
	 *
	 * Can be either next post link or previous.
	 *
	 * @param string $where SQL WHERE clause generated by WordPress
	 */
	public static function modify_nextprevious_post_where( $where ) {
		// The incoming WHERE statement generated by WordPress is a condition for the
		// date, relative to the current post's date. To find the post we want, we
		// just need to get rid of that condition (which is the first) and retain the
		// others.
		if ( self::$loop_navigation_find ) {
			$where = preg_replace( '/WHERE (.+) AND/imsU', 'WHERE', $where );
			c2c_LoopPostNavigationLinks::$loop_navigation_find = false;
		}

		return $where;
	}

} // end c2c_LoopPostNavigationLinks

add_action( 'plugins_loaded', array( 'c2c_LoopPostNavigationLinks', 'init' ) );

endif; // end if !class_exists()



/*
 *
 * TEMPLATE FUNCTIONS
 *
 */


if ( ! function_exists( 'c2c_get_next_or_loop_post_link' ) ) :
/**
 * Gets next post link that is adjacent to the current post, or if none, then
 * the first post in the series.
 *
 * @since 2.5
 *
 * @param string       $format         Optional. Link anchor format. Default is '%link &raquo;'.
 * @param string       $link           Optional. Link permalink format. Default is '%title'.
 * @param bool         $in_same_term   Optional. Whether link should be in a same taxonomy term. Default is false.
 * @param array|string $excluded_terms Optional. Array or comma-separated list of excluded term IDs. Default is ''.
 * @param string       $taxonomy       Optional. Taxonomy, if $in_same_term is true. Default 'category'.
 */
function c2c_get_next_or_loop_post_link( $format='%link &raquo;', $link='%title', $in_same_term = false, $excluded_terms = '', $taxonomy = 'category' ) {
	return c2c_get_adjacent_or_loop_post_link( $format, $link, $in_same_term, $excluded_terms, false, $taxonomy );
}
add_action( 'c2c_get_next_or_loop_post_link', 'c2c_get_next_or_loop_post_link', 10, 5 );
endif;


if ( ! function_exists( 'c2c_next_or_loop_post_link' ) ) :
/**
 * Displays next post link that is adjacent to the current post, or if none, then
 * the first post in the series.
 *
 * @since 2.0
 *
 * @param string       $format         Optional. Link anchor format. Default is '%link &raquo;'.
 * @param string       $link           Optional. Link permalink format. Default is '%title'.
 * @param bool         $in_same_term   Optional. Whether link should be in a same taxonomy term. Default is false.
 * @param array|string $excluded_terms Optional. Array or comma-separated list of excluded term IDs. Default is ''.
 * @param string       $taxonomy       Optional. Taxonomy, if $in_same_term is true. Default 'category'.
 */
function c2c_next_or_loop_post_link( $format='%link &raquo;', $link='%title', $in_same_term = false, $excluded_terms = '', $taxonomy = 'category' ) {
	c2c_adjacent_or_loop_post_link( $format, $link, $in_same_term, $excluded_terms, false, $taxonomy );
}
add_action( 'c2c_next_or_loop_post_link', 'c2c_next_or_loop_post_link', 10, 5 );
endif;


if ( ! function_exists( 'c2c_get_previous_or_loop_post_link' ) ) :
/**
 * Gets previous post link that is adjacent to the current post, or if none,
 * then the last post in the series.
 *
 * @since 2.5
 *
 * @param string       $format         Optional. Link anchor format. Default is '&laquo; %link'.
 * @param string       $link           Optional. Link permalink format. Default is '%title'.
 * @param bool         $in_same_term   Optional. Whether link should be in a same taxonomy term. Default is false.
 * @param array|string $excluded_terms Optional. Array or comma-separated list of excluded term IDs. Default is ''.
 * @param string       $taxonomy       Optional. Taxonomy, if $in_same_term is true. Default 'category'.
 */
function c2c_get_previous_or_loop_post_link( $format='&laquo; %link', $link='%title', $in_same_term = false, $excluded_terms = '', $taxonomy = 'category' ) {
	return c2c_get_adjacent_or_loop_post_link( $format, $link, $in_same_term, $excluded_terms, true, $taxonomy );
}
add_action( 'c2c_previous_or_loop_post_link', 'c2c_previous_or_loop_post_link', 10, 5 );
endif;


if ( ! function_exists( 'c2c_previous_or_loop_post_link' ) ) :
/**
 * Display previous post link that is adjacent to the current post, or if none,
 * then the last post in the series.
 *
 * @since 2.0
 *
 * @param string       $format         Optional. Link anchor format. Default is '&laquo; %link'.
 * @param string       $link           Optional. Link permalink format. Default is '%title'.
 * @param bool         $in_same_term   Optional. Whether link should be in a same taxonomy term. Default is false.
 * @param array|string $excluded_terms Optional. Array or comma-separated list of excluded term IDs. Default is ''.
 * @param string       $taxonomy       Optional. Taxonomy, if $in_same_term is true. Default 'category'.
 */
function c2c_previous_or_loop_post_link( $format='&laquo; %link', $link='%title', $in_same_term = false, $excluded_terms = '', $taxonomy = 'category' ) {
	c2c_adjacent_or_loop_post_link( $format, $link, $in_same_term, $excluded_terms, true, $taxonomy );
}
add_action( 'c2c_previous_or_loop_post_link', 'c2c_previous_or_loop_post_link', 10, 5 );
endif;


if ( ! function_exists( 'c2c_get_adjacent_or_loop_post_link' ) ) :
/**
 * Gets adjacent post link or the post link for the post at the opposite end of the series.
 *
 * Can be either next post link or previous.
 *
 * @since 2.5
 *
 * @param string       $format         Link anchor format.
 * @param string       $link           Link permalink format.
 * @param bool         $in_same_term   Optional. Whether link should be in a same taxonomy term. Default is false.
 * @param array|string $excluded_terms Optional. Array or comma-separated list of excluded term IDs. Default is ''.
 * @param bool         $previous       Optional. Display link to previous post? Default is true.
 * @param string       $taxonomy       Optional. Taxonomy, if $in_same_term is true. Default 'category'.
 */
function c2c_get_adjacent_or_loop_post_link( $format, $link, $in_same_term = false, $excluded_terms = '', $previous = true, $taxonomy = 'category' ) {
	$post = c2c_get_adjacent_or_loop_post( $in_same_term, $excluded_terms, $previous, $taxonomy );

	// Most of what follows was lifted from `get_adjacent_post_link()` (except
	// for passing `$output` through a couple additional filters).

	if ( ! $post ) {
		$output = '';
	} else {
		$title = $post->post_title;

		if ( empty( $post->post_title ) ) {
			$title = $previous ? __( 'Previous Post' ) : __( 'Next Post' );
		}

		/** This filter is documented in wp-includes/post-template.php */
		$title = apply_filters( 'the_title', $title, $post->ID );

		$date = mysql2date( get_option( 'date_format' ), $post->post_date );
		$rel = $previous ? 'prev' : 'next';

		$string = '<a href="' . esc_url( get_permalink( $post ) ) . '" rel="' . $rel . '">';
		$inlink = str_replace( '%title', $title, $link );
		$inlink = str_replace( '%date', $date, $inlink );
		$inlink = $string . $inlink . '</a>';

		$output = str_replace( '%link', $inlink, $format );
	}

	$adjacent = $previous ? 'previous' : 'next';

	/** This filter is documented in wp-includes/link-template.php */
	$output = apply_filters( "{$adjacent}_post_link", $output, $format, $link, $post, $adjacent );

	/**
	 * Filters the adjacent post link.
	 *
	 * The dynamic portion of the hook name, `$adjacent`, refers to the type
	 * of adjacency, 'next' or 'previous'.
	 *
	 * Equivalent to WP's '{$adjacent}_post_link' filter.
	 *
	 * @since 1.5
	 * @since 3.0 Added the `$adjacent` parameter.
	 * @deprecated 2.0 Use 'c2c_{$adjacent}_or_loop_post_link_get'
	 *
	 * @param string  $output   The adjacent post link.
	 * @param string  $format   Link anchor format.
	 * @param string  $link     Link permalink format.
	 * @param WP_Post $post     The adjacent post.
	 * @param string  $adjacent Whether the post is previous or next.
	 */
	$output = apply_filters_deprecated( "{$adjacent}_or_loop_post_link", array( $output, $format, $link, $post, $adjacent ), '2.0', "c2c_{$adjacent}_or_loop_post_link_get" );

	/**
	 * Filters the adjacent post link.
	 *
	 * The dynamic portion of the hook name, `$adjacent`, refers to the type
	 * of adjacency, 'next' or 'previous'.
	 *
	 * Similar to WP's '{$adjacent}_post_link' filter.
	 *
	 * @since 2.5
	 * @since 3.0 Added the `$adjacent` parameter.
	 *
	 * @param string       $output         The adjacent post link.
	 * @param string       $format         Link anchor format.
	 * @param string       $link           Link permalink format.
	 * @param WP_Post      $post           The adjacent post.
	 * @param bool         $in_same_term   Whether link should be in a same taxonomy term. Default is false.
	 * @param array|string $excluded_terms Array or comma-separated list of excluded term IDs. Default is ''.
	 * @param bool         $previous       Whether to display link to previous or next post. Default is true.
	 * @param string       $taxonomy       Taxonomy, if $in_same_term is true. Default 'category'.
	 * @param string       $adjacent       Whether the post is previous or next.
	 */
	return apply_filters( "c2c_{$adjacent}_or_loop_post_link_get", $output, $format, $link, $post, $in_same_term, $excluded_terms, $taxonomy, $adjacent );
}
add_action( 'c2c_get_adjacent_or_loop_post_link', 'c2c_get_adjacent_or_loop_post_link', 10, 6 );
endif;


if ( ! function_exists( 'c2c_adjacent_or_loop_post_link' ) ) :
/**
 * Displays adjacent post link or the post link for the post at the opposite end of the series.
 *
 * Can be either next post link or previous.
 *
 * @param string       $format         Link anchor format.
 * @param string       $link           Link permalink format.
 * @param bool         $in_same_term   Optional. Whether link should be in a same taxonomy term. Default is false.
 * @param array|string $excluded_terms Optional. Array or comma-separated list of excluded term IDs. Default is ''.
 * @param bool         $previous       Optional. Whether to display link to previous or next post. Default is true.
 * @param string       $taxonomy       Optional. Taxonomy, if $in_same_term is true. Default 'category'.
 */
function c2c_adjacent_or_loop_post_link( $format, $link, $in_same_term = false, $excluded_terms = '', $previous = true, $taxonomy = 'category' ) {
	$adjacent = $previous ? 'previous' : 'next';

	/**
	 * Filters the adjacent post link or the post link for the post at the
	 * opposite end of the series.
	 *
	 * @since 2.0
	 * @since 3.0 Added the `$adjacent` parameter and removed `$post` parameter.
	 *
	 * @param string       $format         Link anchor format.
	 * @param string       $link           Link permalink format.
	 * @param bool         $in_same_term   Optional. Whether link should be in a same taxonomy term. Default is false.
	 * @param array|string $excluded_terms Optional. Array or comma-separated list of excluded term IDs. Default is ''.
	 * @param bool         $previous       Optional. Whether to display link to previous or next post. Default is true.
	 * @param string       $taxonomy       Optional. Taxonomy, if $in_same_term is true. Default 'category'.
	 * @param string       $adjacent       Whether the post is previous or next.
	 */
	echo apply_filters(
		"c2c_{$adjacent}_or_loop_post_link_output",
		c2c_get_adjacent_or_loop_post_link( $format, $link, $in_same_term, $excluded_terms, $previous, $taxonomy ),
		$format,
		$link,
		$in_same_term,
		$excluded_terms,
		$previous,
		$taxonomy,
		$adjacent
	);
}
add_action( 'c2c_adjacent_or_loop_post_link', 'c2c_adjacent_or_loop_post_link', 10, 6 );
endif;

if ( ! function_exists( 'c2c_get_adjacent_or_loop_post' ) ) :
/**
 * Returns adjacent post or the post at the opposite end of the series.
 *
 * Can be either next post or previous post.
 *
 * @since 2.6
 *
 * @param bool          $in_same_term   Optional. Whether link should be in a same taxonomy term. Default is false.
 * @param array|string  $excluded_terms Optional. Array or comma-separated list of excluded term IDs. Default is ''.
 * @param bool          $previous       Optional. Whether to display link to previous or next post. Default is true.
 * @param string        $taxonomy       Optional. Taxonomy, if $in_same_term is true. Default 'category'.
 *
 * @return WP_Post|null The post, or null if the post loops back to itself.
 */
function c2c_get_adjacent_or_loop_post( $in_same_term = false, $excluded_terms = '', $previous = true, $taxonomy = 'category' ) {
	if ( $previous && is_attachment() ) {
		$post = get_post( get_post()->post_parent );
	} else {
		$post = get_adjacent_post( $in_same_term, $excluded_terms, $previous, $taxonomy );
	}

	if ( ! $post ) {
		c2c_LoopPostNavigationLinks::$loop_navigation_find = true;
		$post = get_adjacent_post( $in_same_term, $excluded_terms, $previous, $taxonomy );
		// Don't loop to itself.
		if ( $post == get_post() ) {
			$post = null;
		}
	}

	return $post;
}
add_action( 'c2c_get_adjacent_or_loop_post', 'c2c_get_adjacent_or_loop_post', 10, 4 );
endif;

if ( ! function_exists( 'c2c_get_next_or_loop_post' ) ) :
/**
 * Returns next post or the post at the beginning of the series.
 *
 * @since 2.6
 *
 * @param bool          $in_same_term   Optional. Whether link should be in a same taxonomy term. Default is false.
 * @param array|string  $excluded_terms Optional. Array or comma-separated list of excluded term IDs. Default is ''.
 * @param string        $taxonomy       Optional. Taxonomy, if $in_same_term is true. Default 'category'.
 *
 * @return WP_Post|null The post, or null if the post loops back to itself.
 */
function c2c_get_next_or_loop_post( $in_same_term = false, $excluded_terms = '', $taxonomy = 'category' ) {
	return c2c_get_adjacent_or_loop_post( $in_same_term, $excluded_terms, false, $taxonomy );
}
add_action( 'c2c_get_next_or_loop_post', 'c2c_get_next_or_loop_post', 10, 3 );
endif;

if ( ! function_exists( 'c2c_get_previous_or_loop_post' ) ) :
/**
 * Returns previous post or the post at the end of the series.
 *
 * @since 2.6
 *
 * @param bool          $in_same_term   Optional. Whether link should be in a same taxonomy term. Default is false.
 * @param array|string  $excluded_terms Optional. Array or comma-separated list of excluded term IDs. Default is ''.
 * @param string        $taxonomy       Optional. Taxonomy, if $in_same_term is true. Default 'category'.
 *
 * @return WP_Post|null The post, or null if the post loops back to itself.
 */
function c2c_get_previous_or_loop_post( $in_same_term = false, $excluded_terms = '', $taxonomy = 'category' ) {
	return c2c_get_adjacent_or_loop_post( $in_same_term, $excluded_terms, true, $taxonomy );
}
add_action( 'c2c_get_previous_or_loop_post', 'c2c_get_previous_or_loop_post', 10, 3 );
endif;

if ( ! function_exists( 'c2c_get_next_or_loop_post_url' ) ) :
/**
 * Returns the URL for the next post or the post at the beginning of the series.
 *
 * @since 3.0
 *
 * @param bool          $in_same_term   Optional. Whether link should be in a same taxonomy term. Default is false.
 * @param array|string  $excluded_terms Optional. Array or comma-separated list of excluded term IDs. Default is ''.
 * @param string        $taxonomy       Optional. Taxonomy, if $in_same_term is true. Default 'category'.
 *
 * @return WP_Post|null The post, or null if the post loops back to itself.
 */
function c2c_get_next_or_loop_post_url( $in_same_term = false, $excluded_terms = '', $taxonomy = 'category' ) {
	$url = '';
	$post = c2c_get_adjacent_or_loop_post( $in_same_term, $excluded_terms, false, $taxonomy );
	if ( $post ) {
		$url = get_permalink( $post );
	}
	return $url;
}
endif;

if ( ! function_exists( 'c2c_get_previous_or_loop_post_url' ) ) :
/**
 * Returns the URL for the previous post or the post at the end of the series.
 *
 * @since 3.0
 *
 * @param bool          $in_same_term   Optional. Whether link should be in a same taxonomy term. Default is false.
 * @param array|string  $excluded_terms Optional. Array or comma-separated list of excluded term IDs. Default is ''.
 * @param string        $taxonomy       Optional. Taxonomy, if $in_same_term is true. Default 'category'.
 *
 * @return WP_Post|null The post, or null if the post loops back to itself.
 */
function c2c_get_previous_or_loop_post_url( $in_same_term = false, $excluded_terms = '', $taxonomy = 'category' ) {
	$url = '';
	$post = c2c_get_adjacent_or_loop_post( $in_same_term, $excluded_terms, true, $taxonomy );
	if ( $post ) {
		$url = get_permalink( $post );
	}
	return $url;
}
endif;


/*****
 * DEPRECATED FUNCTIONS
 *****/

if ( ! function_exists( 'next_or_loop_post_link' ) ) :
	/**
	 * Display next post link that is adjacent to the current post, or if none,
	 * then the first post in the series.
	 *
	 * @since 1.0
	 * @deprecated 2.0 Use c2c_next_or_loop_post_link() instead
	 */
	function next_or_loop_post_link( $format='%link &raquo;', $link='%title', $in_same_term = false, $excluded_terms = '' ) {
		_deprecated_function( 'next_or_loop_post_link', '2.0', 'c2c_next_or_loop_post_link' );
		return c2c_next_or_loop_post_link( $format, $link, $in_same_term, $excluded_terms );
	}
endif;

if ( ! function_exists( 'previous_or_loop_post_link' ) ) :
	/**
	 * Display previous post link that is adjacent to the current post, or if
	 * none, then the last post in the series.
	 *
	 * @since 1.0
	 * @deprecated 2.0 Use c2c_previous_or_loop_post_link() instead
	 */
	function previous_or_loop_post_link( $format='&laquo; %link', $link='%title', $in_same_term = false, $excluded_terms = '' ) {
		_deprecated_function( 'previous_or_loop_post_link', '2.0', 'c2c_previous_or_loop_post_link' );
		return c2c_previous_or_loop_post_link( $format, $link, $in_same_term, $excluded_terms );
	}
endif;

if ( ! function_exists( 'adjacent_or_loop_post_link' ) ) :
	/**
	 * Display previous post link that is adjacent to the current post, or if
	 * none, then the last post in the series.
	 *
	 * @since 1.0
	 * @deprecated 2.0 Use c2c_adjacent_or_loop_post_link() instead
	 */
	function adjacent_or_loop_post_link( $format, $link, $in_same_term = false, $excluded_terms = '', $previous = true ) {
		_deprecated_function( 'adjacent_or_loop_post_link', '2.0', 'c2c_adjacent_or_loop_post_link' );
		return c2c_adjacent_or_loop_post_link( $format, $link, $in_same_term, $excluded_terms, $previous );
	}
endif;
